accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -1461,3 +1461,319 @@ wrapper class is so huge that it is likely to be a major use case for
+`Iterable<dynamic>`. _This, also still, means that `T` is a bottom type,
+`dynamic` or a type which implements `Iterable<S>` for some `S`. That type may
+just also be an extension type._
+
+If `T` implements `Iterable<S>`, let `E` be `S`, otherwise let `E` be `Never` if
+`T` is a bottom type, or `dynamic` if `T` is `dynamic`.
+
+Now rewrite the `for`-`in` loop to:
+
+```dart
+Iterable<E> _$id1 = e;
+Iterator<E> _$id2 = _$id1.iterator;
+while (_$id2.moveNext()) {
+  D x = _$id2.current;
+  {
+    body
+  }  
+}
+```
+
+Like before, any errors that would occur in that program are reported as errors
+in the original `for`-`in` statement.
+
+_**TODO**: Specify this without using a “desugaring” rewrite. Problems like
+these are exactly why rewrites are problematic, they tend to introduce *more*
+information, through the extra syntax or inferred types, than what the original
+program warranted. Also, the rewrite supposedly happens after type inference, so
+we need to either perform type inference on the desugared code, or assign a
+context type and static type to each new expression, because those are
+referenced by the dynamic semantics._ 
+
+##### Consequences to existing code
+
+The only real change is that the type of `_$id1` is not the static type of `e`.
+That means that if there is a class which overrides its iterator to be more
+specific than required, that information is lost.
+
+Example:
+
+```dart
+class VeryInts extends Iterable<num> {
+  Iterator<int> get iterator => <int>[1].iterator;
+}
+void main() {
+  for (int i in VeryInts()) print(i);
+}
+```
+
+With the previous *specification*, the type that `.iterator` should be accessed
+on was `VeryInts`, which means that `_$id2` will get static type
+`Iterator<int>`, and its `.current` is then assignable to `int i`. With the new
+specification, we don’t look at the members of the actual type, instead we
+immediately up-cast to `Iterable<E>` and work with that.
+
+*Luckily*, our implementations already do that: CFE (Dart2js on DartPad):
+
+> ```
+> lib/main.dart:5:12:
+> Error: A value of type 'num' can't be assigned to a variable of type 'int'.
+> for (int i in VeryInts()) print(i);
+>         ^
+> ```
+
+and Analyzer:
+
+> ```
+> error: line 5 • The type 'VeryInts' used in the 'for' loop must implement 'Iterable' with a type argument that can be assigned to 'int'.
+> ```
+
+So, this looks like it might be a spec-only change for the static semantics, to
+make it match the actual implementation, and it’s *likely* that the
+implementations will also do the right thing when faced with extension types.
+They *must* only invoke instance `iterator`, `moveNext` and `current` members.
+
+#### Iterables and synchronous `yield*`
+
+The synchronous `yield*` operator takes an expression which, like `for`/`in`,
+must be assignable to `Iterable<T>`, and it too must use only instance members to
+iterate the operand where it state that it invokes `iterator`, `moveNext` and
+`current`.
```

Comment:
We should probably have the same level of support for a structural extension-or-extension-type based iteration in a `for` loop and in a `yield*` statement.